app/services/notify_service.rb

```diff
@@ -11,126 +11,175 @@ class NotifyService < BaseService
-    blocked   = @recipient.unavailable?
-    blocked ||= from_self? && @notification.type != :poll
+    def new_account?
+      @sender.created_at > NEW_ACCOUNT_THRESHOLD.ago
+    end
 
-    return blocked if message? && from_staff?
+    def override_for_sender?
+      NotificationPermission.exists?(account: @recipient, from_account: @sender)
+    end
+
+    def from_limited?
+      @sender.silenced? && not_following?
+    end
 
-    blocked ||= domain_blocking?
-    blocked ||= @recipient.blocking?(@notification.from_account)
-    blocked ||= @recipient.muting_notifications?(@notification.from_account)
-    blocked ||= hellbanned?
-    blocked ||= optional_non_follower?
-    blocked ||= optional_non_following?
-    blocked ||= optional_non_following_and_direct?
-    blocked ||= conversation_muted?
-    blocked ||= blocked_mention? if @notification.type == :mention
-    blocked
+    def private_mention_not_in_response?
+      @notification.type == :mention && @notification.target_status.direct_visibility? && !response_to_recipient?
+    end
+
+    def response_to_recipient?
+      return false if @notification.target_status.in_reply_to_id.nil?
+
+      Status.count_by_sql([<<-SQL.squish, id: @notification.target_status.in_reply_to_id, recipient_id: @recipient.id, sender_id: @sender.id, depth_limit: 100]).positive?
+        WITH RECURSIVE ancestors(id, in_reply_to_id, mention_id, path, depth) AS (
+            SELECT s.id, s.in_reply_to_id, m.id, ARRAY[s.id], 0
+            FROM statuses s
+            LEFT JOIN mentions m ON m.silent = FALSE AND m.account_id = :sender_id AND m.status_id = s.id
+            WHERE s.id = :id
+          UNION ALL
+            SELECT s.id, s.in_reply_to_id, m.id, st.path || s.id, st.depth + 1
+            FROM ancestors st
+            JOIN statuses s ON s.id = st.in_reply_to_id
+            LEFT JOIN mentions m ON m.silent = FALSE AND m.account_id = :sender_id AND m.status_id = s.id
+            WHERE st.mention_id IS NULL AND NOT s.id = ANY(path) AND st.depth < :depth_limit
+        )
+        SELECT COUNT(*)
+        FROM ancestors st
+        JOIN statuses s ON s.id = st.id
+        WHERE st.mention_id IS NOT NULL AND s.visibility = 3
+      SQL
+    end
   end
 
-  def conversation_muted?
-    if @notification.target_status
-      @recipient.muting_conversation?(@notification.target_status.conversation)
+  def call(recipient, type, activity)
+    return if recipient.user.nil?
+
+    @recipient    = recipient
+    @activity     = activity
+    @notification = Notification.new(account: @recipient, type: type, activity: @activity)
+
+    # For certain conditions we don't need to create a notification at all
+    return if dismiss?
+
+    @notification.filtered = filter?
+    @notification.save!
+
+    # It's possible the underlying activity has been deleted
+    # between the save call and now
+    return if @notification.activity.nil?
+
+    if @notification.filtered?
+      update_notification_request!
```

Comment:
This is only updated when notifications are created, not when they are deleted, which means the counts can get inaccurate or people can end up with “ghost” notification requests.